### PR TITLE
Launcher and command spi

### PR DESCRIPTION
**Pull request by cescoffier**
PR providing the new Launcher class and the Command SPI. It also implements the `run`, `bare` and `version` commands.

### CODE REVIEW

**Comment by purplefox** [comment]
I took a look at this with Clement in Procida and I was pretty much happy with it.

Julien Paulo @vietj @pmlopes maybe you guys would like to look too?

**Comment by vietj** [comment]
I will have a look soon :-)

File: src/main/java/io/vertx/core/Starter.java [hunk 1]
```diff
@@ -51,7 +51,9 @@
  * {@code java -jar myapp.jar}
  *
  * @author <a href="http://tfox.org">Tim Fox</a>
+ * @deprecated USe {@link Launcher} instead
```
Comment (vietj) [inline src/main/java/io/vertx/core/Starter.java]: minor typo

File: src/main/java/io/vertx/core/impl/ContextImpl.java [hunk 2]
```diff
@@ -240,7 +241,9 @@ public JsonObject config() {
 
   @Override
   public List<String> processArgs() {
-    return Starter.PROCESS_ARGS;
+    // As we are maintaining the launcher and starter class, choose the right one.
+    final List<String> processArgument = VertxCommandLauncher.getProcessArguments();
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/ContextImpl.java]: why final ?

File: src/main/java/io/vertx/core/impl/launcher/CommandLineUtils.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ */
+public class CommandLineUtils {
+
+  /**
+   * @return the fat-jar file used to execute the application if the fat-jar approach is used.
+   */
+  public static String getJar() {
+    // Check whether or not the "sun.java.command" system property is defined,
+    // if it is, check whether the first segment of the command ends with ".jar".
+    final String command = getCommand();
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/launcher/CommandLineUtils.java]: I think you can use `getFirstSegmentOfCommand ()` in the same class to achieve the same thing.

File: src/main/java/io/vertx/core/impl/launcher/VertxLifecycleHooks.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * Interface that let sub-classes of launcher to be notified on different events.
+ *
+ * @author Clement Escoffier <clement@apache.org>
+ */
+public interface VertxLifecycleHooks {
+
+  /**
+   * Hook for sub-classes of the starter class before the vertx instance is started.
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/launcher/VertxLifecycleHooks.java]: you should perhaps mention if it is safe or not to mutate the options in the doc.

File: src/main/java/io/vertx/core/impl/launcher/commands/ClasspathHandler.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      manager = newInstance();
+    }
+
+    final ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(classloader);
+      Method method = manager.getClass().getMethod("deploy", String.class, Vertx.class, DeploymentOptions.class,
+          Handler.class);
+      method.invoke(manager, verticle, vertx, options, completionHandler);
+    } catch (Exception e) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/launcher/commands/ClasspathHandler.java]: perhaps handle InvocationException specially to unwrap and log the cause directly

File: src/main/java/io/vertx/core/spi/launcher/CommandFactoryLookup.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * @author Clement Escoffier <clement@apache.org>
+ */
+public interface CommandFactoryLookup {
+
+  /**
+   * Looks for command implementation and instantiated them.
+   *
+   * @return the set of commands, empty if none are found.
+   */
+  Collection<CommandFactory> lookup();
```
Comment (vietj) [inline src/main/java/io/vertx/core/spi/launcher/CommandFactoryLookup.java]: should be `Collection<CommandFactory<?>>`

File: src/test/java/io/vertx/core/impl/launcher/commands/CommandTestBase.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  protected void waitUntil(BooleanSupplier supplier, long timeout) {
+    long start = System.currentTimeMillis();
+    while (true) {
+      if (supplier.getAsBoolean()) {
+        break;
+      }
+      try {
+        Thread.sleep(10);
+      } catch (InterruptedException ignore) {
+      }
```
Comment (vietj) [inline src/test/java/io/vertx/core/impl/launcher/commands/CommandTestBase.java]: should reset interrupted status with Thread.currentThread().interrupt()

File: src/test/java/io/vertx/test/core/StarterTest.java [hunk 8]
```diff
@@ -57,7 +69,7 @@ public void tearDown() throws Exception {
 
   @Test
   public void testVersion() throws Exception {
-    String[] args = new String[] {"-version"};
+    String[] args = new String[]{"-version"};
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/StarterTest.java]: no need for new String[], same for others below

**Comment by cescoffier** [comment]
Code review integrated.

@vietj

**Comment by vietj** [comment]
will merge it soon.

**Comment by purplefox** [comment]
BTW - if you haven't done so already, don't forget to update the vertx command line section of the docs with the new syntax :)

**Comment by cescoffier** [comment]
Don't worry, the documentation will be updated when all commands are integrated (soon ;-)). The "main-class" used in the `vertx` command and the examples needs to be updated too. 

BTW the syntax didn't change, in the sense it supports the previous way to pass parameters.

**Comment by vietj** [comment]
and write CLI api documentation :-)

**Comment by cescoffier** [comment]
CLI documentation will be provided. I've a first version. I need to check the translation to the other languages.

**Comment by vietj** [comment]
keep in mind polyglot version is needed for the shell.

**Pull request merged**